contracts/liquid-collateral/src/contract.rs

```diff
@@ -0,0 +1,1089 @@
+        },
+        pool_id: msg.pool_id,
+        position_created_address: None,
+        position_id: None,
+        principal_denom: msg.principal_denom,
+        counterparty_denom: msg.counterparty_denom,
+        initial_principal_amount: Uint128::zero(),
+        initial_counterparty_amount: Uint128::zero(),
+        liquidity_shares: None,
+        liquidator_address: None,
+        round_end_time: _env.block.time.plus_seconds(msg.round_duration),
+        auction_duration: msg.auction_duration,
+        auction_end_time: None,
+        auction_principal_deposited: Uint128::zero(),
+        principal_funds_owner: deps.api.addr_validate(&msg.principal_funds_owner)?,
+        principal_to_replenish: Uint128::zero(),
+        counterparty_to_give: None,
+        position_rewards: None,
+        principal_first: msg.principal_first,
+    };
+    STATE.save(deps.storage, &state)?;
+
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
+        .add_attribute("owner", info.sender)
+        .add_attribute("pool_id", msg.pool_id.to_string()))
+}
+
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::CreatePosition(msg) => create_position(deps, env, info, msg),
+        ExecuteMsg::Liquidate => liquidate(deps, env, info),
+        ExecuteMsg::EndRound => end_round(deps, env, info),
+        ExecuteMsg::EndRoundBid(msg) => end_round_bid(deps, env, info, msg),
+        ExecuteMsg::WithdrawBid => withdraw_bid(deps, env, info),
+        ExecuteMsg::ResolveAuction => resolve_auction(deps, env, info),
+    }
+}
+
+#[entry_point]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        // Handle the GetState query
+        QueryMsg::State {} => to_json_binary(&query_get_state(deps)?),
+        QueryMsg::Bid { address } => to_json_binary(&query_bid(deps, address)?),
+        QueryMsg::Bids {} => to_json_binary(&query_get_bids(deps)?),
+        QueryMsg::SortedBids {} => to_json_binary(&query_sorted_bids(deps)?),
+    }
+}
+/*
+
+## Create position with reply
+ - based on the passed position arguments, method is sending MsgCreatePosition to the cl module on Osmosis
+ - in the reply contract is updating the state with the position information.
+*/
+pub fn create_position(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: CreatePositionMsg,
+) -> Result<Response, ContractError> {
+    let mut state = STATE.load(deps.storage)?;
+
+    // Check if the position_id already exists
+    if state.position_id.is_some() {
+        return Err(ContractError::PositionAlreadyExists {});
+    }
+
+    // Only owner can create position - if present
+    match &state.project_owner {
+        Some(owner) if info.sender != *owner => {
+            return Err(ContractError::Unauthorized {});
+        }
+        _ => {}
+    }
+
+    /*
```

Comment:
this is a very useful comment for example, since it states something which otherwise is not obvious